File: libgomp/testsuite/libgomp.oacc-c-c++-common/deep-copy-2.c
#include <assert.h>
#include <stdlib.h>

int
main(int argc, char* argv[])
{
  struct foo {
    int *a, *b, c, d, *e;
  } s;

  s.a = (int *) malloc (16 * sizeof (int));
  s.b = (int *) malloc (16 * sizeof (int));
  s.e = (int *) malloc (16 * sizeof (int));

  #pragma acc data copy(s)
  {
    #pragma acc data copy(s.a[0:10])
    {
      #pragma acc parallel loop attach(s.a)
      for (int i = 0; i < 10; i++)
	s.a[i] = i;
    }
  }

  for (int i = 0; i < 10; i++)
    assert (s.a[i] == i);

  return 0;
}

File: gcc/testsuite/gfortran.dg/goacc/deep-copy-2.f90
! { dg-do compile }

! Test of attach/detach with "acc data", two clauses at once.

program dtype
  implicit none
  integer, parameter :: n = 512
  type mytype
    integer, allocatable :: a(:)
  end type mytype
  integer i

  type(mytype) :: var

  allocate(var%a(1:n))

!$acc data copy(var) copy(var%a) ! { dg-error "Symbol .var. has mixed component and non-component accesses" }

!$acc data copy(var%a) copy(var) ! { dg-error "Symbol .var. has mixed component and non-component accesses" }

!$acc parallel loop
  do i = 1,n
    var%a(i) = i
  end do
!$acc end parallel loop

!$acc end data

!$acc end data

  do i = 1,n
    if (i .ne. var%a(i)) stop 1
  end do

  deallocate(var%a)

end program dtype
